services/management/node.go

```diff
@@ -55,179 +49,78 @@ func NewNodeService(db *reform.DB, registry registry) *NodeService {
 }
 
 // Register do registration of the new node.
-func (s *NodeService) Register(ctx context.Context, req *managementpb.RegisterNodeRequest) (res *managementpb.RegisterNodeResponse, err error) {
-	res = new(managementpb.RegisterNodeResponse)
+func (s *NodeService) Register(ctx context.Context, req *managementpb.RegisterNodeRequest) (*managementpb.RegisterNodeResponse, error) {
+	res := new(managementpb.RegisterNodeResponse)
 
 	if e := s.db.InTransaction(func(tx *reform.TX) error {
-		node, err := s.findNodeByName(tx.Querier, req.NodeName)
-		switch err {
-		case nil:
-			params := &models.UpdateNodeParams{
-				Address:      req.Address,
-				MachineID:    req.MachineId,
-				CustomLabels: req.CustomLabels,
-			}
-			node, err = models.UpdateNode(tx.Querier, node.NodeID, params)
-			if err != nil {
-				return err
-			}
-		case errNodeNotFound:
-			node, err = s.createNewNode(tx.Querier, req)
-			if err != nil {
-				return err
+		node, err := models.FindNodeByName(tx.Querier, req.NodeName)
+		switch status.Code(err) {
+		case codes.OK:
+			if !req.Reregister {
+				return status.Errorf(codes.AlreadyExists, "Node with name %q already exists.", req.NodeName)
 			}
-		default:
-			return err
+			err = models.RemoveNode(tx.Querier, node.NodeID, models.RemoveCascade)
```

Comment:
So if we register new node with the same node, it will be all new node, right?

Reply:
Only if `reregister` is passed. https://jira.percona.com/browse/PMM-4007 (yes, several tickets combined in the single PR 😞)